Add sub-pixel glyph positioning support.

bruvzg opened the pull request:
*Note: This has nothing to do with LCD sub-pixel rendering, it's a completely unrelated technique.*

Instead of rasterizing each glyph vector outline once, it's rasterized multiple times, offset by a fraction of a pixel.

Pros:
- Sharper and a bit more compact text.

Cons:
- Up to 2x (or 4x) more memory used by font (and time spent rasterizing glyphs).

Rendering example (half and quarter modes looks almost identically, but there are slight differences visible on the second screenshot):
<img width="492" alt="rendering" src="https://user-images.githubusercontent.com/7645683/153285782-55555986-bc13-4da4-b64a-762e1d41d81e.png">

5x zoomed part of example (red arrow shows one of the differences between half and quarter modes).
<img width="727" alt="zoom" src="https://user-images.githubusercontent.com/7645683/153285787-182ddbcf-0d88-49de-9d7b-a1e9670744b3.png">

Animated difference between half and quarter modes:
![1](https://user-images.githubusercontent.com/7645683/153291405-a0114d29-9872-4ff5-8574-273225dad5c5.gif)

Related proposal - https://github.com/godotengine/godot-proposals/issues/1258 (probably described issue is mostly caused by hinting not fully applied without sub-pixel positioning, but there's also LCD sub-pixel rendering discussion in the same proposal).

bruvzg committed:
Add sub-pixel glyph positioning support.

AlexFolland commented:
Disclaimer: I am not a contributor to Godot (yet), so please don't consider this comment as authoritative in any way.  I just want to leave my opinion as a Godot user who is subscribed to https://github.com/godotengine/godot-proposals/issues/1258.

I applaud your effort in implementing something new, and especially the level of clarity and detail in your demonstration and explanation.  This was very easy to understand and does add something to consider as an improvement to the engine.

That being said, this does not appeal to me personally.  I wouldn't be enabling it in my projects because it introduces no visual improvement in my personal opinion while introducing a performance detriment.

Even if someone does understand it and see a benefit, it does not resolve the issue in https://github.com/godotengine/godot-proposals/issues/1258, while simultaneously introducing additional complexity in configuration.  Without the same level of clarity as is present here in this MR also present in the documentation, the related option may be difficult to understand for users.

My personal opinion is that LCD subpixel font rendering is the ultimate goal, and anything short of that does not solve the problem, as the text still looks relatively blurry.

Calinou commented:
Could you take a screenshot of the project manager and editor with and without this PR?

bruvzg commented:
> Could you take a screenshot of the project manager and editor with and without this feature enabled (at 100% display scale)?

<details>
  <summary>Project Manager</summary>
  
Disabled:
<img width="508" alt="pm_0" src="https://user-images.githubusercontent.com/7645683/153362024-680560c6-1f2e-4888-8525-1686859a00ea.png">
1/4:
<img width="508" alt="pm_q" src="https://user-images.githubusercontent.com/7645683/153362036-95bcecfa-8904-466e-aa6c-b7c463e71c03.png">
![pm](https://user-images.githubusercontent.com/7645683/153362222-b43ad7df-d520-44e0-8c71-50d08979e6f1.gif)

</details>

<details>
  <summary>Editor</summary>

Disabled:
<img width="793" alt="ed_0" src="https://user-images.githubusercontent.com/7645683/153362136-bda2c70f-d9d8-47ad-9d04-1dfdfaeeecf5.png">
1/4:
<img width="793" alt="ed_q" src="https://user-images.githubusercontent.com/7645683/153362181-3851826f-c2a9-4d82-b65a-7c02f94ea1a2.png">
![ed](https://user-images.githubusercontent.com/7645683/153362101-83342b9b-a85f-4d5a-b955-e495e66d6239.gif)

</details>

I have added `interface/editor/font_subpixel_positioning` editor setting (disabled by default) to change it (can be changed without restarting).

> Up to 2x (or 4x) more memory used by font

This might sound scary, but font cache is not large in the first place, for example real memory impact for the editor: with Chinese localization, and after browsing multiple help pages and scripts to rasterize as much glyph as possible additional 3 MB of RAM were used (2.4 → 5.6), with English localization it's about 0.75 MB increase. No performance impact is noticeable.

Calinou commented:
> No performance impact is noticeable.

My only concern is that new glyphs may cause stuttering, especially on mobile where fonts are often displayed at large sizes.

I suppose enabling ASCII prerendering by default would alleviate this.

Calinou commented:
> but it's disabled by default.

It might be worth enabling this in the editor by default, since I doubt most users will know how to improve font rendering in the editor. The performance impact is probably unnoticeable in the editor.

In projects, I'm not sure if this should be enabled by default. If we also enable ASCII prerendering by default, then it should be fine. Otherwise, I'm worried about stalling during gameplay.

bruvzg commented:
Added font size `AUTO` mode (`1/4` for >= 16, `1/2` for >= 20), and set is default option.

Also added project settings for the default font config (positioning, hinting, aa):

<img width="732" alt="Screenshot 2022-02-12 at 17 49 23" src="https://user-images.githubusercontent.com/7645683/153718153-04a7007d-b07b-44bf-8bf6-0ea45c50c643.png">

mhilbrunner reviewed (approved):
Looks good.

mhilbrunner commented on doc/classes/FontData.xml:
```diff
@@ -600,5 +600,8 @@
 			Font style name.
 		</member>
+		<member name="subpixel_positioning" type="int" setter="set_subpixel_positioning" getter="get_subpixel_positioning" enum="TextServer.SubpixelPositioning" default="1">
+			Font glyph sub-pixel positioning mode.
```
Could we document _why_ you would use this? Basically that it trades performance for sharper rendering, especially for smaller font sizes, and may not have much effect for large font sizes.

We could also recommend the AUTO setting and point to the project setting.

akien-mga commented:
Thanks!

Calinou commented:
We should probably see if there's a point in exposing the editor setting. To me, it seems there's little reason to change it from the default, as the Auto setting makes a good job at only using subpixel positioning when it makes a noticeable difference.

AlexFolland commented:
If you ever find yourself making a subjective argument for what value a setting should be set to, that's when you know the setting needs to be exposed, as others may disagree.
